### PR TITLE
Tall building check with tag validation

### LINKED ISSUE #535: TallBuildingCheck
**Description**
The purpose of this check is to identify invalid building:levels and height tags as well as identify statistically outlying building:levels/height tags.

**Requirements**
Check Candidate 
 * Is an `Edge`|`Relation`|`Area`|
 * Is a building or part of a building
 * Has building:levels or height tag

**Use Cases**
Case 1 - Building has a building:levels tag greater than 100.
* Small amount of these flags but this was the ask for the original purpose of this check. 

Case 2 - Building has an invalid building:levels tag.
* The building (house) [id:795395318](https://www.openstreetmap.org/way/795395318) has an invalid building:levels tag. (should only be comprised of numerical values)
image

Case 3 - Building has an outlying building:levels tag compared to buildings around it.
* The building [id:734691356](https://www.openstreetmap.org/way/734691356) has an outlying height tag compared to it's neighboring homes.
image

Case 4 - Building has an invalid height tag.
* The building [id:788365480](https://www.openstreetmap.org/way/788365480) has an invalid height tag. (needs space between '2' and 'm')
image

Case 5 - Building has an outlying height tag compared to buildings surrounding it.
* The building (house) [id:710607756](https://www.openstreetmap.org/way/710607756) has an outlying height tag compared to it's neighboring homes.
image

**Supported regions**
All countries

**References**
[OSM Wiki for Height Tag](https://wiki.openstreetmap.org/wiki/Key:height#:~:text=Examples%20%20%20%20Height%20%20%20,Use%20a%20typewriter%20apostrophe%20character%20af%20...%20)
[OSM Wiki for Building:Levels Tag](https://wiki.openstreetmap.org/wiki/Key:building:levels)

### PR DESCRIPTION
**Pull request by reichg**
### Description:

***This check has 5 cases:***

1) building:levels tag > 100
2) building:levels tag is invalid (syntax/cannot be parsed)
3) building:levels tag is an outlier compared to it's neighboring buildings
4) height tag is invalid (syntax/cannot be parsed/contains invalid characters)
5) height tag is an outlier compared to it's neighboring buildings

### Test Results:

***Case 1***
building:levels tag > 100
|   ISO | Sampled  | TP | FP | False Positive Rate |
|--------|----------|-----|----|-------------------- |
|  AUS   |    0        |   0   |   0 |          0%                 |
|  BRA   |    0        |   0   |   0 |          0%                |
|  COL   |    0        |   0   |   0 |          0%                 |
|  FRA   |    0        |   0   |   0 |          0%                 |
|  NOR   |    0        |   0   |   0 |          0%                 |
|  PER  |    0        |   0   |   0 |          0%                 |
|  USA   |    0        |   0   |   0 |          0%                 |

***Case 2***
building:levels tag is invalid (syntax/cannot be parsed)
|   ISO | Sampled  | TP | FP | False Positive Rate |
|--------|----------|-----|----|-------------------- |
|  AUS   |    178       |   178   |   0 |          0%                 |
|  BRA   |    68        |   0   |   0 |          0%                 |
|  COL   |    17        |   16   |   1 |          5.8%                 |
|  FRA   |    7        |   7   |   0 |          0%                 |
|  NOR   |    0        |   0   |   0 |          0%                 |
|  PER  |    6        |   6   |   0 |          0%                 |
|  USA   |    126        |   10   |   0 |          0%                 |

***Case 3***
building:levels tag is an outlier compared to it's neighboring buildings
|   ISO | Sampled  | TP | FP | False Positive Rate |
|--------|----------|-----|----|-------------------- |
|  AUS   |    154        |   154   |   0 |          0%                 |
|  BRA   |    65        |   66   |   0 |          0%                 |
|  COL   |    115    |   116 |   0 |          0%                 |
|  FRA   |    75        |   73   |   2 |          2.667%                 |
|  NOR   |    19        |   19   |   0 |          0%                 |
|  PER  |    6        |   6   |   0 |          0%                 |
|  USA   |    150        |   150   |   0 |          0%                 |

***Case 4***
height tag is invalid (syntax/cannot be parsed/contains invalid characters)
|   ISO | Sampled  | TP | FP | False Positive Rate |
|--------|----------|-----|----|-------------------- |
|  AUS   |    57        |   57   |   0 |          0%                 |
|  BRA   |    300        |   300   |   0 |          0%                 |
|  COL   |    10        |   10   |   0 |          0%                 |
|  FRA   |    142        |   142   |   0 |          0%                 |
|  NOR   |    6        |   6   |   0 |          0%                 |
|  PER  |    29        |   29   |   0 |          0%                 |
|  USA   |    200        |   200   |   0 |          0%                 |

***Case 5***
height tag is an outlier compared to it's neighboring buildings
Average FPR: 4.6%
|   ISO | Sampled  | TP | FP | False Positive Rate |
|--------|----------|-----|----|-------------------- |
|  AUS   |    29        |   29   |   0 |          0%                 |
|  BRA   |    12        |   8   |   4 |          33.3%                 |
|  COL   |    6        |   3   |   3 |          50%                 |
|  FRA   |    15        |   14   |   1 |          6.667%                 |
|  NOR   |    1        |   1   |   0 |          0%                 |
|  PER  |    1        |   1   |   0 |          0%                 |
|  USA   |    127        |   126   |   1 |          0.079%                 |

Resolves #535

### CODE REVIEW

File: config/configuration.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
       "org.openstreetmap.atlas.checks.validation"
     ],
     "type": "org.openstreetmap.atlas.checks.base.BaseCheck",
-    "enabled.value.default": true
+    "enabled.value.default": false
```
Comment (sayas01) [inline config/configuration.json:7]: Please revert this change

File: config/configuration.json [hunk 2]
```diff
@@ -1079,6 +1079,27 @@
...
     "minHighwayType": "tertiary"
   },
+  "TallBuildingCheck": {
+    "enabled": true,
+    "bufferDistanceMeters": 1600.0,
+    "minDatasetSizeForStatsComputation": 50.0,
+    "maxLevelTagValue": 100.0,
+    "outlierMultiplier": 18.0,
+    "invalidHeightCharacters": ["~","`","!","@","#","$","%","^","&","*","(", ")","-","_","+","=","[","[","}","]","|","\\",":",";","<",",",">","?","/"],
+    "magicNumbers": {
```
Comment (sayas01) [inline config/configuration.json:1089]: Since these are just instruction indices, I don't think we need to make these configurable values.

File: config/configuration.json [hunk 3]
```diff
@@ -1079,6 +1079,27 @@
     },
     "minHighwayType": "tertiary"
   },
+  "TallBuildingCheck": {
+    "enabled": true,
+    "bufferDistanceMeters": 1600.0,
+    "minDatasetSizeForStatsComputation": 50.0,
+    "maxLevelTagValue": 100.0,
+    "outlierMultiplier": 18.0,
+    "invalidHeightCharacters": ["~","`","!","@","#","$","%","^","&","*","(", ")","-","_","+","=","[","[","}","]","|","\\",":",";","<",",",">","?","/"],
```
Comment (sayas01) [inline config/configuration.json:1088]: Please refer to the naming convention used throughout the file for the config keys.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+        this.outlierMultiplier = this.configurationValue(configuration, "outlierMultiplier",
+                OUTLIER_MULTIPLIER_DEFAULT);
+        this.invalidHeightCharacters = new HashSet<>(this.configurationValue(configuration,
+                "invalidHeightCharacters", INVALID_CHARACTER_DEFAULT));
+        this.magicNumber3 = this.configurationValue(configuration, "magicNumbers.3", THREE_DEFAULT);
+        this.magicNumber4 = this.configurationValue(configuration, "magicNumbers.4", FOUR_DEFAULT);
+        this.magicNumberOneQuarter = this.configurationValue(configuration,
+                "magicNumbers.oneQuarter", ONE_QUARTER_DEFAULT);
+        this.magicNumberThreeQuarters = this.configurationValue(configuration,
+                "magicNumbers.threeQuarters", THREE_QUARTER_DEFAULT);
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:96]: Please refer to the naming convention used throughout the file for the config keys.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+     * This is the actual function that will check to see whether the object needs to be flagged.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        markAsFlagged(object.getOsmIdentifier());
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:127]: Looks like the OsmIdentifier is being marked as flagged and the atlasIdentifier is checked if flagged in validCheckForObject.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+        if (this.hasInvalidHeightTag(heightTag))
+        {
+            return Optional.of(this.createFlag(object, this
+                    .getLocalizedInstruction((int) this.magicNumber4, object.getOsmIdentifier())));
+        }
+
+        // Don't run statistics if building contains building=apartments tag (apartments seem to be
+        // showing up a lot as outliers)
+        if (!(tags.containsKey(BuildingTag.KEY) && tags.get(BuildingTag.KEY)
+                .equals(BuildingTag.APARTMENTS.toString().toLowerCase())))
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:530]: Same comment as before, could replace equals() with equals equalsIgnoreCase().

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 7]
```diff
@@ -0,0 +1,728 @@
...
+     * @return boolean if it has specific tag.
+     */
+    private boolean hasHeightTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(HeightTag.KEY);
+    }
+
+    private boolean hasInvalidHeightTag(final String heightTag)
+    {
+        return (!heightTag.contains(" ") && heightTag.contains("m"))
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:480]: What is the reasoning for `(!heightTag.contains(" ") && heightTag.contains("m"))`

**Review by sayas01 — COMMENTED** [review]
@reichg Love the breakdown of analysis based on various use cases. Although the FP rate for case 5 is very high in BRA and COL, since the average FP rate is less than 10%, this should be good to go once comments are addressed. Kudos to you for a very well documented code💯.  My only other suggestion apart from the existing comments would be to improve the test coverage.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 8]
```diff
@@ -0,0 +1,728 @@
...
+                    return this.getStatisticsWithNewStoreEntry(object, buildingLevelsTagValue,
+                            BuildingLevelsTag.KEY);
+                }
+
+            }
+
+        }
+
+        // Case 2: Building has an invalid building:levels tag because it cannot be parsed.
+        catch (final Exception e)
```
Comment (seancoulter) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:217]: could you try wrapping the exception only around the parseDouble call? Just to limit the scope in case another, unrelated exception occurs somewhere else in the try{} block and this catches it

**Review by seancoulter — COMMENTED** [review]
Nice work! This is an interesting check. Do you have an example of one of the FPs in COL and BRA? I'd just like to see if there's a pattern causing that high FP rate that can be programmatically addressed.

File: config/configuration.json (hunk 1 above)
Comment (reichg) [reply config/configuration.json:7]: yeah sorry this happened because I had to test the output locally one last tie and forgot to revert.

**Comment by reichg** [comment]
> @reichg Love the breakdown of analysis based on various use cases. Although the FP rate for case 5 is very high in BRA and COL, since the average FP rate is less than 10%, this should be good to go once comments are addressed. Kudos to you for a very well documented code💯. My only other suggestion apart from the existing comments would be to improve the test coverage.

yeah it was a struggle to even get the test coverage above 50% since I would need to create a test atlas with at least 50 buildings to have coverage over the statistical part of the check. I just wanted to get this passing above 50 before I attempt adding the tests for those statistical cases. :-P

File: config/configuration.json (hunk 3 above)
Comment (reichg) [reply config/configuration.json:1088]: I am not sure I know what you mean by this, please explain when you get the chance!

Comment (reichg) [reply config/configuration.json:1088]: bufferDistanceMeters: -> buffer.distance.meters: ?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 7 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:480]: this is a specific height tag case that needs to be modeled with with a numerical value, then a space, then a unit of measurement according to [Osm Wiki](https://wiki.openstreetmap.org/wiki/Key:height) for height tags. I am just capturing incorrectly modeled tags such as -> "5m". That tag should be -> "5 m".

**Comment by reichg** [comment]
> Nice work! This is an interesting check. Do you have an example of one of the FPs in COL and BRA? I'd just like to see if there's a pattern causing that high FP rate that can be programmatically addressed.

Yeah I found that in some countries apartment buildings will show up as outliers because they're significantly taller than nearby buildings so I removed apartments from the statistical part of this check. Those FPs will be lower.

File: config/configuration.json (hunk 1 above)
Comment (reichg) [reply config/configuration.json:7]: addressed

File: config/configuration.json [hunk 9]
```diff
@@ -1079,6 +1079,27 @@
     },
     "minHighwayType": "tertiary"
   },
+  "TallBuildingCheck": {
+    "enabled": true,
```
[the remark this answers is not in the document]
Comment (reichg) [reply config/configuration.json:1083]: addressed

File: config/configuration.json (hunk 2 above)
Comment (reichg) [reply config/configuration.json:1089]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 10]
```diff
@@ -0,0 +1,728 @@
...
+import org.openstreetmap.atlas.tags.HeightTag;
+import org.openstreetmap.atlas.tags.RelationTypeTag;
+import org.openstreetmap.atlas.tags.annotations.validation.Validators;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+import org.openstreetmap.atlas.utilities.scalars.Distance;
+
+import groovy.lang.Tuple4;
+
+/**
+ * Auto generated Check template
```
[the remark this answers is not in the document]
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:32]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 11]
```diff
@@ -0,0 +1,728 @@
...
+            if (buildingLevelsTagValue > this.maxLevelTagValue)
+            {
+                return Optional.of(this.createFlag(object,
+                        this.getLocalizedInstruction(0, object.getOsmIdentifier())));
+            }
+
+            // Don't run statistics if building contains building=apartments tag (apartments seem to
+            // be showing up a lot as outliers)
+            if (!(tags.containsKey(BuildingTag.KEY) && tags.get(BuildingTag.KEY)
+                    .equals(BuildingTag.APARTMENTS.toString().toLowerCase())))
```
[the remark this answers is not in the document]
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:185]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 5 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:127]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 6 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:530]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 8 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:217]: addressed

**Comment by sayas01** [comment]
> > @reichg Love the breakdown of analysis based on various use cases. Although the FP rate for case 5 is very high in BRA and COL, since the average FP rate is less than 10%, this should be good to go once comments are addressed. Kudos to you for a very well documented code💯. My only other suggestion apart from the existing comments would be to improve the test coverage.
> 
> yeah it was a struggle to even get the test coverage above 50% since I would need to create a test atlas with at least 50 buildings to have coverage over the statistical part of the check. I just wanted to get this passing above 50 before I attempt adding the tests for those statistical cases. :-P

@reichg To improve test coverage for the statistical cases, you could leverage the [loadFromTextResource()](https://github.com/osmlab/atlas/blob/88a19a8e796e721c358310a13f4e345e471f7935/src/main/java/org/openstreetmap/atlas/utilities/testing/TestAtlas.java#L336) method to load a test atlas with the specific use case instead of creating a large test atlas. MalformedPolylineCheckTest class has an example.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 7 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:483]: @reichg Got it! Is there possibility of other unit of measurement usage for height tag for different countries?

**Comment by reichg** [comment]
> > > @reichg Love the breakdown of analysis based on various use cases. Although the FP rate for case 5 is very high in BRA and COL, since the average FP rate is less than 10%, this should be good to go once comments are addressed. Kudos to you for a very well documented code💯. My only other suggestion apart from the existing comments would be to improve the test coverage.
> > 
> > 
> > yeah it was a struggle to even get the test coverage above 50% since I would need to create a test atlas with at least 50 buildings to have coverage over the statistical part of the check. I just wanted to get this passing above 50 before I attempt adding the tests for those statistical cases. :-P
> 
> @reichg To improve test coverage for the statistical cases, you could leverage the [loadFromTextResource()](https://github.com/osmlab/atlas/blob/88a19a8e796e721c358310a13f4e345e471f7935/src/main/java/org/openstreetmap/atlas/utilities/testing/TestAtlas.java#L336) method to load a test atlas with the specific use case instead of creating a large test atlas. MalformedPolylineCheckTest class has an example.

Thinking I will take a break from this check as I have spent a lot of time on it. I can come back to it though to increase the coverage if that is okay.

File: config/configuration.json (hunk 3 above)
Comment (sayas01) [reply config/configuration.json]: Yp, the changes look good!

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 7 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:483]: Technically yes but meters is the default and takes up about 95% of the values that belong to the key.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 12]
```diff
@@ -0,0 +1,723 @@
...
+        }
+
+        // Don't run statistics if building contains building=apartments tag (apartments seem to
+        // be showing up a lot as outliers)
+        if (!(tags.containsKey(BuildingTag.KEY)
+                && tags.get(BuildingTag.KEY).equalsIgnoreCase(BuildingTag.APARTMENTS.toString())))
+        {
+            // Check store to see if building intersects stored entry rectangle
+            final Optional<Tuple4<String, Double, Double, Double>> rectangleStatsWhichIntersectObjectOptional = this
+                    .getRectangleStatsWhichIntersectObject(this.storedAreasWithStatistics, object,
```
Comment (seancoulter) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:188]: nit: `this.storedAreasWithStatistics` shouldn't need to be a param here

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 13]
```diff
@@ -0,0 +1,723 @@
...
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        markAsFlagged(object.getOsmIdentifier());
+        final Map<String, String> tags = object.getOsmTags();
+
+        // level tag logic
+        if (this.hasBuildingLevelTag(tags))
```
Comment (seancoulter) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:122]: any undesirable outcome if the object has a buildingLevel tag and a height tag?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java [hunk 14]
```diff
@@ -0,0 +1,723 @@
...
+                    // compared to surrounding buildings with "height tag.
+                    if (rectangleStatsWhichIntersectObjectOptional.isEmpty())
+                    {
+                        return this.getStatisticsWithNewStoreEntry(object,
+                                buildingHeightTagValue.get(), HeightTag.KEY);
+                    }
+                }
+
+            }
+            catch (final Exception ignored)
```
Comment (seancoulter) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:561]: what can be thrown here? Can we use a more specific Exception subclass if possible and log the error?

**Review by seancoulter — COMMENTED** [review]
Just some smaller things I noticed. Last big thing I'm thinking of right now is runtime - how does this check look in terms of that?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 12 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:188]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 13 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:122]: the first scenario that the feature fits under, if any, short circuits the flag method even if the feature might fit under other scenarios. Should I try to catch all scenarios with the flag?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 14 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:561]: Actually can probably remove this try/catch section since the parsing double try/catch is within the parseHeightTag function.

**Comment by reichg** [comment]
> Just some smaller things I noticed. Last big thing I'm thinking of right now is runtime - how does this check look in terms of that?

AUS completion time -> 7.593 minutes

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java (hunk 13 above)
Comment (seancoulter) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/TallBuildingCheck.java:122]: addressed offline

**Review by seancoulter — APPROVED** [review]
I think this is looking good. Thanks for addressing all the update requests. AUS runtime looks fine but in case we run into longer times say in building-dense shards, we can play around with decreasing the buffer distance

**Comment by reichg** [comment]
> I think this is looking good. Thanks for addressing all the update requests. AUS runtime looks fine but in case we run into longer times say in building-dense shards, we can play around with decreasing the buffer distance

Keep in mind that is local runtime with legacy sparkjob

**Review by sayas01 — APPROVED** [review]
Nice work @reichg! Thanks for addressing all the comments!
